builtin/logical/transit/path_decrypt.go

```diff
@@ -44,36 +71,78 @@ func (b *backend) pathDecrypt() *framework.Path {
 
 func (b *backend) pathDecryptWrite(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	name := d.Get("name").(string)
-	ciphertext := d.Get("ciphertext").(string)
-	if len(ciphertext) == 0 {
-		return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
-	}
-
+	batchInputRaw := d.Get("batch").(string)
+	var batchInputItems []BatchRequestItem
+	var batchInput []byte
 	var err error
-
-	// Decode the context if any
-	contextRaw := d.Get("context").(string)
-	var context []byte
-	if len(contextRaw) != 0 {
-		context, err = base64.StdEncoding.DecodeString(contextRaw)
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
 		if err != nil {
-			return logical.ErrorResponse("failed to base64-decode context"), logical.ErrInvalidRequest
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+
+		if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputItems); err != nil {
+			return nil, fmt.Errorf("invalid input: %v", err)
+		}
+
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
+
+		batchInputItems = make([]BatchRequestItem, 1)
+		batchInputItems[0] = BatchRequestItem{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
 		}
 	}
 
-	// Decode the nonce if any
-	nonceRaw := d.Get("nonce").(string)
-	var nonce []byte
-	if len(nonceRaw) != 0 {
-		nonce, err = base64.StdEncoding.DecodeString(nonceRaw)
-		if err != nil {
-			return logical.ErrorResponse("failed to base64-decode nonce"), logical.ErrInvalidRequest
+	batchResponseItems := make([]BatchResponseItem, len(batchInputItems))
+
+	var contextSet bool
+	if len(batchInputItems) == 0 {
+		return logical.ErrorResponse("missing input to process"), logical.ErrInvalidRequest
+	}
+
+	contextSet = batchInputItems[0].Context != ""
+
+	for i, item := range batchInputItems {
+		// Invalidate the items that are not expected in the request item, for
+		// safety
+		batchInputItems[i].Plaintext = ""
+		batchInputItems[i].DecodedContext = nil
+		batchInputItems[i].DecodedNonce = nil
```

Comment:
I think a better way to handle this would be to get rid of the DecodedContext/DecodedNonce struct members entirely and make Context/Nonce into []byte. The Go JSON decoder will automatically de-base64 those values on ingress. That gets rid of a bunch of boilerplate below, makes the three above statements unnecessary (although they are anyways), and simplifies the struct.